Add splitter for Markdown, taking document hierarchy into account

kabir opened the pull request:
<!--
Thank you so much for your contribution!

Please fill in all the sections below.
Please open the PR as a draft initially. Once it is reviewed and approved, we will ask you to add documentation and examples.
Please note that PRs with breaking changes or without tests will be rejected.

Please note that PRs will be reviewed based on the priority of the issues they address.
We ask for your patience. We are doing our best to review your PR as quickly as possible.
Please refrain from pinging and asking when it will be reviewed. Thank you for understanding!
-->

## Issue
<!-- Please specify the ID of the issue this PR is addressing. For example: "Closes #1234" or "Fixes #1234" -->
Closes #2417

## Change
<!-- Please describe the changes you made. -->
I added a MarkdownSectionSplitter to split Markdown documents by headers.

The header goes into the metadata of the resulting TextSegments, while the body of the section goes into the TextSegments.

The Markdown must be well-formed, we don't try to do any validation beyond what the user gives us.

The list of TextSegments list the sections in order, while the metadata of each segment allows us to find where in the document the segment came from.

Headers appearing in code blocks are ignored (some languages use '#' for comments). 

A user may optionally specify a splitter to further split the sections.

Finally, there is a hook to override the creation of the Document  used to further split each section. This allows the caller to add of more metadata for each section.


## General checklist
<!-- Please double-check the following points and mark them like this: [X] -->
- [X] There are no breaking changes
- [X] I have added unit and/or integration tests for my change
- [C] The tests cover both positive and negative cases
- [X] I have manually run all the unit and integration tests in the module I have added/changed, and they are all green
- [X] I have manually run all the unit and integration tests in the [core](https://github.com/langchain4j/langchain4j/tree/main/langchain4j-core) and [main](https://github.com/langchain4j/langchain4j/tree/main/langchain4j) modules, and they are all green
<!-- Before adding documentation and example(s) (below), please wait until the PR is reviewed and approved. -->
- [ ] I have added/updated the [documentation](https://github.com/langchain4j/langchain4j/tree/main/docs/docs)
- [ ] I have added an example in the [examples repo](https://github.com/langchain4j/langchain4j-examples) (only for "big" features)
- [ ] I have added/updated [Spring Boot starter(s)](https://github.com/langchain4j/langchain4j-spring) (if applicable)


## Checklist for adding new maven module
<!-- Please double-check the following points and mark them like this: [X] -->
- [ ] I have added my new module in the root `pom.xml` and `langchain4j-bom/pom.xml`


## Checklist for adding new embedding store integration
<!-- Please double-check the following points and mark them like this: [X] -->
- [ ] I have added a `{NameOfIntegration}EmbeddingStoreIT` that extends from either `EmbeddingStoreIT` or `EmbeddingStoreWithFilteringIT`
- [ ] I have added a `{NameOfIntegration}EmbeddingStoreRemovalIT` that extends from `EmbeddingStoreWithRemovalIT`

## Checklist for changing existing embedding store integration
<!-- Please double-check the following points and mark them like this: [X] -->
- [ ] I have manually verified that the `{NameOfIntegration}EmbeddingStore` works correctly with the data persisted using the latest released version of LangChain4j

kabir commented on langchain4j/src/test/java/dev/langchain4j/data/document/splitter/MarkdownSplitterTest.java:
```diff
@@ -0,0 +1,250 @@
+        checkTextSegment(source, segments.get(5), "Section 2.1.2", "Section 2.1", 3, 1, "section 2.1.2\nmore");
+        checkTextSegment(source, segments.get(6), "Section 2.2", "Section 2", 2, 1, ".");
+        checkTextSegment(source, segments.get(7), "Section 2.2.1", "Section 2.2", 3, 0, ".");
+        checkTextSegment(source, segments.get(8), "Section 3", "Title", 1, 2, "section 3");
```
I have one doubt here. Here the level is taken to be 2, inferred from the number of '#' in the header.

However, in this case there is no level 1 parent since I knowingly 'broke' the structure of the document a bit. So the path to this section is:

\# Title
\### Section 3

elsewhere in the tests I am following the more standard structure, e.g:

\# Title
\## Section 2
\### Section 2.1

So my question is whether 2 for the level here is ok, or if I should use the 'effective' level which is 1.

glaforge commented on langchain4j/src/main/java/dev/langchain4j/data/document/splitter/MarkdownSectionSplitter.java:
```diff
@@ -0,0 +1,339 @@
+    public static final String SECTION_HEADER = "md-section-header";
+    public static final String SECTION_INDEX_WITHIN_PARENT = "md-section-index-in-parent";
+    public static final String SECTION_PARENT_HEADER = "md-parent-header";
+    private static final Pattern HEADER_PATTERN = Pattern.compile("^#+ .*");
```
In Markdown, sometimes people also use headers underlined with 3+ `=` signes and 3+ `-` signs.
It's maybe a bit harder to parse with a regex though, but maybe we should not forget about those headers too.

kabir replied:
I gave this a go. 
3+ = seems to result in h1, no matter how many '=' I put.
3+ - seems to result in h2, no matter how many '-' I put.

Perhaps an option is if I document in the splitter the expectations for the headers, and then they could use a DocumentTransformer before splitting to massage the markdown into the expected format?

Here is my example, in a code block first, and then the same markdown rendered below

```
Normal:
# h1
## h2
### h3
### h4


3+ === seems to result in h1

h1
===
h1
====
h1
=====
h1
======
h1
=======


3+ ---- seems to result in h2

h2
---
h2
----
h2
-----
h2
------
```

Normal:
# h1
## h2
### h3
### h4


3+ === seems to result in h1

h1
===
h1
====
h1
=====
h1
======
h1
=======


3+ ---- seems to result in h2

h2
---
h2
----
h2
-----
h2
------

glaforge commented on langchain4j/src/main/java/dev/langchain4j/data/document/splitter/MarkdownSectionSplitter.java:
```diff
@@ -0,0 +1,339 @@
+    }
+
+    @Override
+    public List<TextSegment> split(Document document) {
```
General comment, but the hook could be added here, probably:

There are lots of Markdown documents that contain some "front-matter" preamble, with various metadata information. It's usually written as yaml, between two triple dashes lines.
It would be interesting to have such metadata added to the metadata of the document.

kabir replied:
Thanks @glaforge 

Do you have some examples of the front-matter preamble please?

kabir replied:
Never mind, I can see how it works.
If the first line is minimum three dashes, it is the pre-amble, and is hidden. If I don't close it, everything in the document becomes 'hidden'. 

```
------
hello
---
Start doc
```

If there is any content before the pre-amble it is no longer hidden.

```
<!-- test -->
------
hello
-----
Start doc
```

So it seems I should check if the document starts with a line which only contains  dashes, and there are at least three. And then a similar line further down closes this. 

I'll look into adding something so the user can parse this.

kabir replied:
I see this also has the same effect and hides the 'hello':

```
+++
hello
+++
Start doc
```
I also tried with *, and a few other characters.

Again, I think it is reasonable to specify the format we expect, so the user can fix up any weirdness with a transformer?

glaforge replied:
Here's a [pointer](https://www.markdownguide.org/basic-syntax/#alternate-syntax) that shows the alternative header syntax. Triple stars, or triple dashes, can be used standalone as [horizontal bars](https://www.markdownguide.org/basic-syntax/#horizontal-rules).

For the [front-matter](https://daily-dev-tips.com/posts/what-exactly-is-frontmatter/), I think it's not part of the Markdown but I've seen that frequently used in Markdown files. For example, the static blog generator I use ([Hugo](https://gohugo.io/content-management/front-matter/)) uses that for specifying tags, and other meta attributes for the document.

> Again, I think it is reasonable to specify the format we expect, so the user can fix up any weirdness with a transformer?

Unfortunately, I have to disagree here. Developers will send Markdown coming from various places, from libraries that convert HTML to Markdown, or other kind of documents, and they don't have control over how they write their Markdown. So we can't really mandate the exact format the Markdown should have, and we have to make it work with whatever they send us.

Also, I'm wondering if we shouldn't use a Markdown parsing library that would understand all the subtleties of Markdown, like [flexmark](https://github.com/vsch/flexmark-java) or [commonmark-java](https://github.com/commonmark/commonmark-java). I'm curious if you've looked at those? 

I also had a question (or maybe a feature request) regarding the composability. 
Somehow, the other splitters we have can be composed hierarchically, so I'm wondering if you're considering doing that for your splitter too: ie. I may want to first split by Markdown headers, but then do some sentence splitting for example.

Another aspect is the configuration and depth of parsing: for example, I may want to chunk to get sections under 2000 characters and/or tokens, so if a section is under that size, no need to split at a smaller header level. I don't think it's possible right now to do that, but was curious if you were considering this too.

kabir replied:
I vaguely looked at flexmark and commonmark, but that was for our conversion from html to markdown. Flexmark seems quite unmaintained, and generated very bad markdown from the html, so I ended up using Docling and forgot all about those libraries.. 

But it could be a good idea to use commonmark to do the parsing. I'll investigate that in the project I am working on, which is what inspired contributing the splitter.

Another thing I realised I don't like in this PR is the builder. It would probably be easier to just do:

```
MarkdownSectionSplitter splitter = new MarkdownSectionSplitter();
```
and then if something extra needs setting to just do
```
splitter.setX(x);
```

Regarding the composability question, I don't totally get what you are asking :-) I am passing in another splitter which further splits each section once we've extracted the headers. I am doing that in `testSectionSplitter()`. Do you mean something else?

Regarding the last question, that is something I have not thought about. Our use-case is currently to extract all sub-sections, so that is what this does. So, I don't think that is something I'll add at this stage

glaforge replied:
Regarding builders, yes, if there's just one or two parameters, it can be overkill (and verbose).

For the composability, sorry that I wasn't very clear.
I'm thinking of recursive splitters.
For example, we have `DocumentSplitters.recursive(...)` which does this: it chains different splitters, like paragraph splitter, then if the chunk is still too big, it chunks by sentence, then by words, and characters. So the splitting goes deeper if the split still gives too big chunks.

So instead of manually splitting once with the Markdown header splitter, then splitting again by sentence or something else, you just chain them together.

dliubarskyi replied:
I see @glaforge already mentioned markdown parsing libraries, I completely agree, we should use them if possible. 

In general, it would be very helpful to have one or few examples of markdown files with multiple sections and subsections and demonstrate how you expect it to be split (given a few `maxSegmentSize` settings).

dliubarskyi replied:
Regarding the builder: we keep our objects immutable unless there is a good reason not to. So builder is more preferable than setters

glaforge reviewed (commented):
Comment about alternative syntax for headers, and the treatment of front-matter metadata.

dliubarskyi commented:
@kabir thank you! I did not look into the implementation yet, but I see there are many discussions about the formatting and all the variations. Just wondering if we should use some library to parse markdown into AST. It seems there are a few java libs for that. WDYT?

kabir commented:
Thanks, I'll try to look into all this, although it might take some time!

kabir committed:
Add splitter that can split markdown by sections.

Note that this currently needs a SNAPSHOT of commonmark-java
containing https://github.com/commonmark/commonmark-java/pull/361.

kabir committed:
Add temporary replacement for MarkdownRenderer

This includes the change needed in
https://github.com/commonmark/commonmark-java/pull/361

and removes the previous snapshot

kabir commented:
Hi @glaforge and @dliubarskyi I have updated the pull request to use Commonmark to do the parsing.

Originally, I tried to implement my own visitor, but ran into lots of problems trying to output the original markdown from the parts I am not doing any special handling for. So I found the rendererer I am using now, which made it a lot easier. Unfortunately, the renderer has a amall bug which I am trying to fix in https://github.com/commonmark/commonmark-java/pull/361. I've not heard back on the pull request, so I have added a `TempMarkdownRenderer` which includes the fix, and otherwise delegates to Commonmark as usual.

This splitter now supports our use-case. Assuming, what I have put in is ok, may I suggest that for any improvements you do this as follow-up work?

@glaforge suggested something about chained splitters, which I **think** I am doing, at least for what we want to achieve. I feel he has something different in mind, which isn't what we want to do. But I may well be misunderstanding something.

kabir committed:
Headers aren't always plain text.

e.g. we have some in our code that are xml elements

fabriziofortino commented:
@kabir any update on this PR? Happy to help!

kabir commented:
@fabriziofortino Hi, my focus has shifted elsewhere since February. Feel free to take it over, I still think it can be useful to others :-)
